spec.bs

```diff
@@ -452,6 +457,93 @@ To <dfn>build an interest group permissions url</dfn> given a [=origin=] |ownerO
+  * <code>10\*1024\*1024</code> total [=interest group/estimated size=] of [=interest groups=] per
+    [=interest group/owner=].
+
+<div algorithm>
+To <dfn>perform storage maintenance</dfn>:
+
+1. Let |ownersAndExpiry| be a new [=ordered map=] whose [=map/keys=] are [=origins=] and
+  [=map/values=] are [=moments=].
+
+  Note: The [=map/key=] is from [=interest group/owner=], and [=map/value=] is from
+    [=interest group/expiry=].
+
+1. Let |now| be the [=current wall time=].
+1. [=list/For each=] |ig| of the [=user agent=]'s [=interest group set=]:
+  1. Let |owner| be |ig|'s [=interest group/owner=].
+  1. If |ig|'s [=interest group/expiry=] is before |now|, then [=list/remove=] |ig| from the
+    [=user agent=]'s [=interest group set=] and [=iteration/continue=].
+  1. If |ownersAndExpiry|[|owner|] [=map/exists=], then [=map/set=] |ownersAndExpiry|[|owner|] to
+    |ig|'s [=interest group/expiry=] if it comes after |ownersAndExpiry|[|owner|].
+  1. Otherwise, [=map/set=] |ownersAndExpiry|[|owner|] to |ig|'s [=interest group/expiry=].
+1. If |ownersAndExpiry|'s [=map/size=] &gt; 2000, then [=map/set=] |ownersAndExpiry| to
+  |ownersAndExpiry| [=map/sorted in descending order=] with |a| being less than |b| if |a|'s
+  [=map/value=] comes before |b|'s [=map/value=], where [=map/values=] are [=interest group/expiry=].
+
+  Note: In order to remove interest groups of owners expiring soonest first, sort owners based on
+    their expiry in descending order, 
+
+1. Let |owners| be the [=map/get the keys|keys=] of |ownersAndExpiry|.
+1. [=list/For each=] |i| in [=the range=] from 0 to |owners|'s [=set/size=], exclusive:
+  1. If |i| &ge; 2000, then [=list/remove=] [=interest groups=] from the [=user agent=]'s
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|], and [=iteration/continue=].
+  1. Let |regularIgs| be a [=list=] of [=regular interest groups=] in the [=user agent=]'s
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If |regularIgs|'s [=list/size=] &gt; 1000, then [=clear excess interest groups=] with
+    |regularIgs|, |owners|[|i|], and 1000.
+  1. Let |negativeIgs| be a [=list=] of [=negative interest groups=] in the [=user agent=]'s
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If |negativeIgs|'s [=list/size=] &gt; 20000, then [=Clear excess interest groups=] with
+    |negativeIgs|, |owners|[|i|], and 20000.
+1. [=list/For each=] |owner| of |owners|:
+  1. Let |igs| be a [=list=] of [=interest groups=] in the [=user agent=]'s [=interest group set=]
+    whose [=interest group/owner=] is |owner|, [=list/sorted in descending order=] with |a| being
+    less than |b| if |a|[=interest group/expiry=] comes before |b|[=interest group/expiry=].
+  1. Let |cumulativeSize| be 0.
+  1. [=list/For each=] |ig| of |igs|:
+    1. If the sum of |cumulativeSize| and |ig|'s [=interest group/estimated size=]

```

Comment:
This needs to be ...

```
    1. Increment |cumulativeSize| by |ig|'s [=interest group/estimated size=].
    1. If the sum of |cumulativeSize| and |ig|'s [=interest group/estimated size=] &gt; <code>10\*1024\*1024</code>, then [=list/remove=] |ig| from the [=user agent=]'s [=interest group set=].
```

The way it's currently written (the "otherwise" optimization) doesn't remove the IG that causes the `cumulativeSize` to exceed the limit.

Reply:
Oh, you're right. I missed "the sum of |cumulativeSize| and |ig|'s [=interest group/estimated size=]" when I first read it. (I'd interpreted it as just |cumulativeSize|.) This is strictly better than what we had before. Thanks.